Print offending token in error message from containsSubsequence

h908714124 opened the pull request:
Fix #2395

Added two lines to the error message from `containsSubsequence`.

Only in `Iterables.containsSubsequence`.

`Arrays.containsSubsequence` is working with `Object` and that would be more work.

New error message looks like this:

```
Failed to find token at subsequence index 3 in actual:
 "Han"
Expecting actual:
  ["Yoda", "Luke", "Leia", "Obi-Wan"]
to contain subsequence:
  ["Luke", "Leia", "Obi-Wan", "Han", "C-3PO", "R2-D2", "Anakin"]
```

h908714124 commented:
I'm going to squash three commits into one here

h908714124 commented:
I did a rebase that should not have changed anything, yet the MacOS build has failed with this message:

```
org.assertj.core.util.Files_fileNamesIn_Test
java.io.IOException: 
	at org.assertj.core.util.Closeables_closeQuietly_Test.should_ignore_thrown_errors(Closeables_closeQuietly_Test.java:37)
```

Looks like an unrelated random test failure to me, re-running the checks should fix it.

scordio commented:
> Looks like an unrelated random test failure to me, re-running the checks should fix it.

No worries about that, we have a few flaky tests around.

h908714124 commented:
The error from the windows test has someting in common with the previous MacOS error. There's probably something wrong with `Closeables_closeQuietly_Test.java:37`. Why is `Closeables_closeQuietly_Test` called from several other tests?

```
org.assertj.core.perf.SoftAssertionsPerfTest
java.io.IOException: 
[INFO] Running org.assertj.core.perf.TypeComparatorsPerfTest
	at org.assertj.core.util.Closeables_closeQuietly_Test.should_ignore_thrown_errors(Closeables_closeQuietly_Test.java:37)
```

joel-costigliola commented:
@h908714124 I'm following your idea but slightly rewording the error message to be like:
```
Expecting actual to contain the specified subsequence but failed to find subsequence element 4 (at index 2) in actual.
actual:
  [1, 2, 1, 3, 5, 1, 2, 3, 3, 6, 7]
subsequence:
  [2, 3, 4]
```
thoughts?

h908714124 commented:
The wording is fine, but the offending token should be on a separate line with no other text before it. Otherwise it might reduce readability, especially for long tokens like in the issue. So how about we slightly shift the line breaks and do this instead:

```
Expecting actual to contain the specified subsequence but failed to find subsequence element 
  "Han" 
(at subsequence index 3) in actual:
  ["Yoda", "Luke", "Leia", "Obi-Wan"]
subsequence:
  ["Luke", "Leia", "Obi-Wan", "Han", "C-3PO", "R2-D2", "Anakin"]
```

h908714124 commented:
Note that the actual formatting I get in my IDE (not when running the test, but when using assertj in practice) has a different formatting with more line breaks. It looks something like this:

```
Expecting actual:
  ["Yoda",
   "Luke",
   "Leia",
   "Obi-Wan"]
to contain subsequence:
  ["Luke",
   "Leia",
   "Obi-Wan",
   "Han",
   "C-3PO",
   "R2-D2",
   "Anakin"]
```

So the actual formatting already contains each token on a separate line.

h908714124 commented:
In the future, we might further improve the formatting, for example by adding a little arrow when printing the subsequence:

```
  ["Luke",
   "Leia",
   "Obi-Wan",
> "Han",
   "C-3PO",
   "R2-D2",
   "Anakin"]
```

h908714124 commented:
Updated PR to print a different message, please check again.
Also added a new test method `should_fail_if_subsequence_is_bigger_than_actual_real_message` that tests the actual error message. The existing tests were not sensitive to changes in the error message.

joel-costigliola commented:
the reason for the line breaks is AssertJ breaking down long iterables, you can control this with https://assertj.github.io/doc/#maxlengthforsinglelinedescription

joel-costigliola commented:
I'm actually writing different error messages when:

1. the subsequence has more elements than actual

```
Expecting actual to contain the specified subsequence but actual does not have enough elements to contain it, actual size is 4 when subsequence size is 7
actual:
  ["Yoda", "Luke", "Leia", "Obi-Wan"]
subsequence:
  ["Luke", "Leia", "Obi-Wan", "Han", "C-3PO", "R2-D2", "Anakin"]
```

1. actual does not have enough elements left to match the rest of the subsequence, examples:

```
Expecting actual to contain the specified subsequence but actual does not have enough elements left to compare after reaching element 3 out of 4 with 3 subsequence element(s) still to find.
actual:
  ["Yoda", "Luke", "Leia", "Obi-Wan"]
subsequence:
  ["Leia", "Obi-Wan", "Han"]
```
or
```
Expecting actual to contain the specified subsequence but actual does not have enough elements left to compare after reaching element 5 out of 5 with 1 subsequence element(s) still to find.
actual:
  ["Leia", "Luke", "Yoda", "Obi-Wan", "Anakin"]
subsequence:
  ["Leia", "Obi-Wan", "Han"]
```

joel-costigliola commented:
@h908714124, I'm taking over this PR as I'm doing some refactoring along the way (but you'll get full credit for it :wink:).  

I'm actually writing different error messages when:

1. the subsequence has more elements than actual

```
Expecting actual to contain the specified subsequence but actual does not have enough elements to contain it, actual size is 4 when subsequence size is 7
actual:
  ["Yoda", "Luke", "Leia", "Obi-Wan"]
subsequence:
  ["Luke", "Leia", "Obi-Wan", "Han", "C-3PO", "R2-D2", "Anakin"]
```

1. if not falling in the previous failure case, we go with the original error message enhanced with your suggestion

```
Expecting actual to contain the specified subsequence but failed to find the element at subsequence index 2 in actual:
subsequence element not found in actual:
  "Han"
actual:
  ["Leia", "Luke", "Yoda", "Obi-Wan", "Anakin"]
subsequence:
  ["Leia", "Obi-Wan", "Han"]
```

joel-costigliola commented:
closed by https://github.com/assertj/assertj/commit/8093f73b9d9d61c37f35092e063d128bfd9e6431
